x/wasm/internal/keeper/keeper.go

```diff
@@ -684,15 +729,134 @@ func (k Keeper) GetByteCode(ctx sdk.Context, codeID uint64) ([]byte, error) {
 	return k.wasmer.GetCode(codeInfo.CodeHash)
 }
 
+func (k Keeper) dispatchAll(ctx sdk.Context, contractAddr sdk.AccAddress, ibcPort string, subMsgs []wasmvmtypes.SubMsg, msgs []wasmvmtypes.CosmosMsg) error {
+	// first dispatch all submessages (and the replies).
+	err := k.dispatchSubmessages(ctx, contractAddr, ibcPort, subMsgs)
+	if err != nil {
+		return err
+	}
+	// then dispatch all the normal messages
+	return k.dispatchMessages(ctx, contractAddr, ibcPort, msgs)
+}
+
 func (k Keeper) dispatchMessages(ctx sdk.Context, contractAddr sdk.AccAddress, ibcPort string, msgs []wasmvmtypes.CosmosMsg) error {
 	for _, msg := range msgs {
-		if err := k.messenger.Dispatch(ctx, contractAddr, ibcPort, msg); err != nil {
+		events, _, err := k.messenger.DispatchMsg(ctx, contractAddr, ibcPort, msg)
+		if err != nil {
+			return err
+		}
+		// redispatch all events, (type sdk.EventTypeMessage will be filtered out in the handler)
+		ctx.EventManager().EmitEvents(events)
+	}
+	return nil
+}
+
+func (k Keeper) dispatchMsgWithGasLimit(ctx sdk.Context, contractAddr sdk.AccAddress, ibcPort string, msg wasmvmtypes.CosmosMsg, gasLimit uint64) (events []sdk.Event, data [][]byte, err error) {
+	// FIXME: do I need to wrap the BlockGasMeter in the same way?
+	limitedMeter := sdk.NewGasMeter(gasLimit)
+	subCtx := ctx.WithGasMeter(limitedMeter)
+
+	// catch out of gas panic and just charge the entire gas limit
+	defer func() {
+		if r := recover(); r != nil {
+			// if it's not an OutOfGas error, raise it again
+			// FIXME: do I need to do something special to get the proper stack trace here?
```

Comment:
We have similar logic in the querier https://github.com/CosmWasm/wasmd/blob/master/x/wasm/internal/keeper/querier.go#L157-L175

Reply:
Okay, that wants to catch all panics, I want to re-raise any other panic (just handle out of gas here).